# Guided Project: Exploring eBay Car Sales Data

The aim of this project is to clean up the eBay car sales dataset and perform basic initial analysis on it.

Question: ?

## 1. Importing Libraries and DataSet



In [2]:
import pandas as pd
import numpy as np

#The line of code below imports the autos.csv into the python environment as a dataframe and specifies the encoding unique to this importation.
autos = pd.read_csv("autos.csv",encoding= "latin1")



In [3]:
#This line of code shows jupyters ability to reflect the properties of dataframes without calling them?
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
#Next up, information on the dataset contained within the dataframe is required.
autos.info()

autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Information on the eBay Car Sales DataSet

From the above information obtained by using the info() method on the autos dataset, the following can 
be deduced:
1. A number of the columns in the dataframe contain null values.
2. The case type used isn't consistent; a combination of CamelCase and Snakecase.
3. The naming convention used for the column headings need to align with python's naming convention.
4. The dataframe contains a combination of object type and integer type.

## 2. Cleaning up column headings in the Dataframe.

In [5]:
#The code below uses the columns attribute to print an array containing the column names.
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
#To edit the column names, a copy of the array is created, worked on, then assigned back to the dataframe.
autos_copy = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
                 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
                 'odometer', 'registration_month', 'fuel_type', 'brand',
                 'unrepaired_damage', 'date_created', 'pictures_number', 'postal_code',
                 'last_seen']

#Assign the copy back to the dataframe
autos.columns = autos_copy
print(autos.columns)

#Examine the current state of the autos dataframe 
autos.head()

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'date_created', 'pictures_number', 'postal_code',
       'last_seen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage         date_created  pictures_number  \
0     peugeot              nein  2016-03-26 00:00:00                0   
1         bmw              nein  2016-04-04 00:00:00                0   
2  volkswagen              nein  2016-03-26 00:00:00                0   
3       smart              nein  2016-03-12 00:00:00                0   
4        ford              nein  2016-04-01 00:00:00                0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

The changes were made above in line with pythons syntax convention for naming to ensure easy access and use.

## 3. DataSet Cleaning

Firstly, we need to determine what cleaning tasks need to be performed on this particular dataset by taking a closer look on the data.

In [7]:
# Two things we can consider when looking out for discrepancies are; repetitive text and numeric data stored as text.

#Exploring the data frame.
autos.describe(include= "all")


date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

               date_created  pictures_number   postal_code  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### Information on the DataSet:

From the information provided from our describe() method, the following notes have been deduced:
1. The "seller" and "offer_type" columns mostly contain the same values hence we can neglect their relevance.
2. The "odometer" and "price" columns need to be converted to acceptable data types.
3. The minimum and maximum values in the "registration_year" column are inaccurate.
4. The "pictures_number" column does not have any relevant data.




In [8]:
# For the Price and Odometer columns, we convert the columns to a numeric dtype and take out the text.
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].astype("int")
print(autos["odometer"])
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype("int")
print(autos["price"])

# Rename the 'odometer' column, and drop the unnecessary columns
autos = autos.rename({'odometer': 'odometer_km'}, axis=1)
print(autos.head)

0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 50000, dtype: int64
0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64
<bound method NDFrame.head of               date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                

/tmp/ipykernel_94/4150788699.py:6: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$","")


In [9]:
#Taking a closer look at the "odometer_km" and "price" columns,checking for the data that we fear doesn't look right.
#First, we check for all the unique values.
print(autos["odometer_km"].unique().shape)

#Next, we check for the min/max/median/mean values in each series.
print(autos["odometer_km"].describe())



(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [10]:
print(autos["price"].unique().shape)

print(autos["price"].describe())

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [11]:
print(autos["odometer_km"].value_counts().head())
print(autos["odometer_km"].sort_index(ascending= True).head())

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64
0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64


In [12]:
print(autos["price"].value_counts().head())
print(autos["price"].sort_index(ascending= True).head())

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64
0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64


## Findings from the Odometer and Price column.
Odometer: A lot of the cars in the dataset have high odometer readings seen from the frequency of 150000 as well as having 13 unique values.

Price: 1421 cars have a price of 0 which is pretty much impossible, hence further inspection is required. The highest price of cars is at $99,999,999, which also seems accurate. We would need to further examine the highest and lowest prices in the column.

In [13]:
#For the lower end of the price column we'd be looking at the last 20 price values.
print(autos["price"].value_counts().sort_index(ascending= True).head(20))

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


From the above, it can be seen that $1 represents the price of a significant number of cars, probably an event type of activity i.e auction is holding for the vehicles considering how low the value is.

In [14]:
print(autos["price"].value_counts().sort_index(ascending= False).head(20))

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


From the above, it can be seen that an abrupt jump is recorded from 350000 to 999990 hence we remove the dataset beneath the 350000 mark.

In [15]:
#Removing the outliers in the price column 
autos = autos[autos["price"].between(1,350000)]

#Displaying how it looks after filtering.
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the Date column

We explore the "date_crawled", "ad_created" and "last_seen" columns to understand the format presented for each date object. Where necessary, we convert to numeric representations to enable a stronger quantitiative knowledge.


In [16]:
#Starting by exploring the first 5 rows of the dataframe to get a general understanding.
print(autos[["date_crawled","date_created","last_seen"]][0:5])

#Next up, we isolate the first 10 string objects in each data point on the date_crawled series.
print(autos["date_crawled"].str[:10])

          date_crawled         date_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object


In [17]:
#Calculating the distribution of values in the "focus" columns as percentages.
#Starting with the "date_crawled" column
print(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64


According to the findings above, the date_crawled column shows a near uniform distribution of dates across a one-month period (March - April).

In [18]:
#"date_created" column
print(autos["date_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: date_created, Length: 76, dtype: float64


The date created column reveals a synonymous distribution across the the first months in the year 2015 with a change occuring in the year 2016. The year 2016 revealed a similar distribution across dates in the month April

In [19]:
#"last_seen" column
print(autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


In [20]:
# "registration_year" column, .describe() fxn since values already exist as numerics.
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

As revealed earlier, the registration_year has a number of incorrect values in its series. The minimum value i.e 1000 shows a year before cars were made and the maximum i.e 9999 is many years into the future.

We need to figure out how to work with this data and avoid inaccuracy in our analysis.

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [21]:
#After deciding the upper and lower bounds to work with i.e 1900 and 2016, we remove values outside those bounds.
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize= True).head(8)


2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
Name: registration_year, dtype: float64

From the findings above, the majority of cars listed in the dataset were registered in the last 20 years i.e 1999 - Present.

## Exploring Price by Brand

Using aggregation, we explore the brand column to reveal unique values to aggregate by, create a storage for the value and calculate the mean of those unique values.

In [22]:
#First off, we explore the "brand" column.
print(autos["brand"].describe())
brands = autos["brand"].value_counts().head(20)

count          46681
unique            40
top       volkswagen
freq            9862
Name: brand, dtype: object


The first twenty (20) cars are to be aggregated against ranging from the top brand i.e volkswagen to the 20th brand i.e mini.

In [23]:
#We select the 20 most common cars as the dataset to work on aggregation with.
most_common_20_cars = brands.head(20).index
print(most_common_20_cars)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')


In [24]:
brand_mean_prices = {}

for brand in most_common_20_cars:
     brand_row = autos[autos["brand"] == brand]
     mean_value = brand_row["price"].mean()
     brand_mean_prices[brand] = mean_value
    
print(brand_mean_prices)
    

{'volkswagen': 5402.410261610221, 'bmw': 8332.820517811953, 'opel': 2975.2419354838707, 'mercedes_benz': 8628.450366422385, 'audi': 9336.687453600594, 'ford': 3749.4695065890287, 'renault': 2474.8646069968195, 'peugeot': 3094.0172290021537, 'fiat': 2813.748538011696, 'seat': 4397.230949589683, 'skoda': 6368.0, 'nissan': 4743.40252454418, 'mazda': 4112.596614950635, 'smart': 3580.2239031770046, 'citroen': 3779.1391437308866, 'toyota': 5167.091062394604, 'hyundai': 5365.254273504273, 'sonstige_autos': 12338.550218340612, 'volvo': 4946.501170960188, 'mini': 10613.459657701711}


From the above mean prices, lets work with the to six (6) brands:
1. The "audi", "BMW" and "Mercedes Benz" brands are typically high priced vehicles.
2. Budget friendly brands include "Ford", "Opel" etc.
3. The "Volkswagen" model exists in between.

## Storing Aggregate Data in a DataFrame

From the mean price analysis, we discovered the distinct price grap in the top 6 vehicle brands. We can use aggregation to understand the average mileage for those cars and connect any possible links. i.e Does more expensive mean more mileage?



In [28]:
#First off, we'll be converting the brand_mean_prices dictionary into a series using the pandas series constructor.
bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

#Then using the pandas Dataframe constructor, we create a single column dataframe with a column title specified.
bmp_df = pd.DataFrame(bmp_series, columns=["mean_price"])
print(bmp_df)

volkswagen         5402.410262
bmw                8332.820518
opel               2975.241935
mercedes_benz      8628.450366
audi               9336.687454
ford               3749.469507
renault            2474.864607
peugeot            3094.017229
fiat               2813.748538
seat               4397.230950
skoda              6368.000000
nissan             4743.402525
mazda              4112.596615
smart              3580.223903
citroen            3779.139144
toyota             5167.091062
hyundai            5365.254274
sonstige_autos    12338.550218
volvo              4946.501171
mini              10613.459658
dtype: float64
                  mean_price
volkswagen       5402.410262
bmw              8332.820518
opel             2975.241935
mercedes_benz    8628.450366
audi             9336.687454
ford             3749.469507
renault          2474.864607
peugeot          3094.017229
fiat             2813.748538
seat             4397.230950
skoda            6368.000000
nissan           

In [31]:
mean_mileage_dictionary = {}
mean_price_dictionary = {}


#Calculating the mean mileagenand price for the top 6 brands and storing in a dictionary.
for brand in most_common_20_cars[:6]:
     brand_row = autos[autos["brand"] == brand]
     mean_value = brand_row["odometer_km"].mean()
     mean_mileage_dictionary[brand] = mean_value
    
for brand in most_common_20_cars[:6]:
     brand_row = autos[autos["brand"] == brand]
     mean_value = brand_row["price"].mean()
     mean_price_dictionary[brand] = mean_value
        
print(mean_mileage_dictionary)
print(mean_price_dictionary)

{'volkswagen': 128707.15879132022, 'bmw': 132572.51313996495, 'opel': 129310.0358422939, 'mercedes_benz': 130788.36331334666, 'audi': 129157.38678544914, 'ford': 124266.01287159056}
{'volkswagen': 5402.410261610221, 'bmw': 8332.820517811953, 'opel': 2975.2419354838707, 'mercedes_benz': 8628.450366422385, 'audi': 9336.687453600594, 'ford': 3749.4695065890287}


In [32]:
#Converting both dictionaries to a series constructor using the aforementioned pandas series constructor.
mmd_series = pd.Series(mean_mileage_dictionary)
print(mmd_series)
mpd_series = pd.Series(mean_price_dictionary)
print(mpd_series)


volkswagen       128707.158791
bmw              132572.513140
opel             129310.035842
mercedes_benz    130788.363313
audi             129157.386785
ford             124266.012872
dtype: float64
volkswagen       5402.410262
bmw              8332.820518
opel             2975.241935
mercedes_benz    8628.450366
audi             9336.687454
ford             3749.469507
dtype: float64


In [42]:
#Creating a dataframe from the mean_mileage series object using the pandas dataframe constructor.
mmd_df = pd.DataFrame(mmd_series, columns=["mean_mileage"])
mmd_df["mean_price"] = mpd_series
print(mmd_df)


                mean_mileage   mean_price
volkswagen     128707.158791  5402.410262
bmw            132572.513140  8332.820518
opel           129310.035842  2975.241935
mercedes_benz  130788.363313  8628.450366
audi           129157.386785  9336.687454
ford           124266.012872  3749.469507


## Conclusion

From the above information, it can be seen that the price of the vehicle does not necessarily tie into the mileage it possesses. Al